components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/RoleBasedScopeIssuer.java

```diff
@@ -0,0 +1,758 @@
+        List<String> userRoleList;
+        if (preservedCaseSensitive) {
+            userRoleList = Arrays.asList(userRoles);
+        } else {
+            userRoleList = new ArrayList<>();
+            for (String aRole : userRoles) {
+                userRoleList.add(aRole.toLowerCase(Locale.ENGLISH));
+            }
+        }
+        // Iterate the requested scopes list.
+        for (String scope : requestedScopes) {
+            boolean isRestrictUnassignedScopes = OAuth2ServiceComponentHolder.isRestrictUnassignedScopes();
+            // Get the set of roles associated with the requested scope.
+            if ((isRestrictUnassignedScopes && oAuthServerConfiguration.getAllowedScopes().contains(scope)) ||
+                    appScopes.containsKey(scope) || !isRestrictUnassignedScopes) {
+                addAuthorizedRoles(appScopes, scope, preservedCaseSensitive, userRoleList, authorizedScopes);
+            }
+        }
+        return (!authorizedScopes.isEmpty()) ? authorizedScopes : defaultScope;
+    }
+
+    private void addAuthorizedRoles(Map<String, String> appScopes, String scope, boolean preservedCaseSensitive,
+                                    List<String> userRoleList, List<String> authorizedScopes) {
+
+        String roles = appScopes.get(scope);
+        // If the scope has been defined in the context of the App and if roles have been defined for the scope.
+        if (roles != null && !roles.isEmpty()) {
+            List<String> roleList = new ArrayList<>();
+            for (String aRole : roles.split(",")) {
+                if (preservedCaseSensitive) {
+                    roleList.add(aRole.trim());
+                } else {
+                    roleList.add(aRole.trim().toLowerCase(Locale.ENGLISH));
+                }
+            }
+            // Check if user has at least one of the roles associated with the scope.
+            roleList.retainAll(userRoleList);
+            if (!roleList.isEmpty()) {
+                authorizedScopes.add(scope);
+            }
+        } else {
+            authorizedScopes.add(scope);
+        }
+    }
+
+    /**
+     * Extract the roles from the user attributes.
+     *
+     * @param userAttributes retrieved from the token
+     * @return roles
+     */
+    private String[] getRolesFromUserAttribute(Map<ClaimMapping, String> userAttributes, String roleClaim) {
+
+        for (Map.Entry<ClaimMapping, String> entry : userAttributes.entrySet()) {
+            if (null != entry.getKey().getLocalClaim()) {
+                if (roleClaim.equals(entry.getKey().getLocalClaim().getClaimUri()) && StringUtils
+                        .isNotBlank(entry.getValue())) {
+                    return entry.getValue().replace("\\/", "/").
+                            replace("[", "").replace("]", "").
+                            replace("\"", "").split(FrameworkUtils.getMultiAttributeSeparator());
+                }
```

Comment:
this belongs in a separate function

Reply:
also let's try to use java8 streams API here. Would improve readability

Reply:
```
userAttributes.entrySet().stream()
            .filter(entry -> entry.getKey().getLocalClaim() != null)
            .filter(entry -> roleClaim.equals(entry.getKey().getLocalClaim().getClaimUri()))
            .map(Map.Entry::getValue)
            .filter(StringUtils::isNotBlank)
. . .
```

Reply:
fixed https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2084/commits/ad343f6bb0ec0f4978d6e599ac94b1835dab76de